[srvls][SRVOCF-278] Adding Python functions docs (#32181)
openshift/openshift-docs: modules/serverless-python-function-return-values.adoc

// Module included in the following assemblies
//
// * /serverless/functions/serverless-developing-python-functions.adoc

[id="serverless-python-function-return-values_{context}"]
= Python function return values

Functions may return any value supported by https://flask.palletsprojects.com/en/1.1.x/quickstart/#about-responses[Flask] because the invocation framework proxies these values directly to the Flask server.

.Example
[source,python]
----
def main(context: Context):
    data = { "message": "Howdy!" }
    headers = { "content-type": "application/json" }

>     return body, 200, headers

Reviewer: I'm not a much familiar with python and Flask api, but I think this will not work since `body` was not defined. 
I tried that I got an error. 

So maybe it would be good to change to:

`body = { "message": "Howdy!" }`
or
`return data, 200, headers`

Author: @lance wdyt?

Reviewer 2: Oh yeah - good catch. Setting `body` is probably the more illustrative approach. `body = { "message": "Howdy!" }`

Author: Adding in the next commit, lmk if it looks correct 🙂

----

> 

Reviewer: I think here we could have just body and keep the return sentence (something like that, was I meant first time):

```
def main(context: Context):
    body = { "message": "Howdy!" }
    headers = { "content-type": "application/json" }
    return body, 200, headers
```

Functions can set both headers and response codes as secondary and tertiary response values from function invocation.

== Returning CloudEvents

Developers can use the `@event` decorator to tell the invoker that the function return value must be converted to a CloudEvent before sending the response.

.Example
[source,python]
----
@event("event_source"="/my/function", "event_type"="my.type")
def main(context):
    # business logic here
    data = do_something()
    # more data processing
    return data
----

> This example sends a CloudEvent as the response value, with a type of `"my.type"`, a source of `"/my/function"`, and the data property set to `data`. The `event_source` and `event_type` decorator attributes are both optional.

Author: @lance on this one does it mean to call out

1. "The `return` property is set to `data`"
2. "The `data` property is set to `do_something()`"
3. something else?

This originally came from the content [here](https://github.com/boson-project/func/blob/main/docs/guides/python.md#cloudevents) so I'm not positive which you meant.

> This example sends a CloudEvent as the response value, with a type of `"my.type"`, a source of `"/my/function"`, and the data property set to `data`. The `event_source` and `event_type` decorator attributes are both optional.

Reviewer 2: @abrennan89 the `data` variable returned from the `main()` function is sent as the `data` property on the CloudEvent response. How about rewording it similar to this, with a link to the CloudEvent spec? Does that help?

```suggestion
This example sends a CloudEvent as the response value, with a type of `"my.type"` and a source of `"/my/function"`. The returned `data` variable is set as the CloudEvent's [`data` property](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#event-data). The `event_source` and `event_type` decorator attributes are both optional.
```

> This example sends a CloudEvent as the response value, with a type of `"my.type"`, a source of `"/my/function"`, and the data property set to `data`. The `event_source` and `event_type` decorator attributes are both optional.

Author: Updated in next commit, PTAL

> If not specified, the CloudEvent `event_source` attribute is set to `"/parliament/function"`, and the `event_type` attribute is set to `"parliament.response"`.

Author: @lance can you provide explanations for the `parliament` ones cos I'm not sure what they do and it's not explained [here](https://github.com/boson-project/func/blob/main/docs/guides/python.md#cloudevents) - thanks!

> If not specified, the CloudEvent `event_source` attribute is set to `"/parliament/function"`, and the `event_type` attribute is set to `"parliament.response"`.

Reviewer 2: @abrennan89 the `event_source` and `event_type` attributes refer to properties in a Python CloudEvent, as manifested by [cloudevents/sdk-python](https://github.com/cloudevents/sdk-python). These properties are required per the CloudEvents spec. For convenience, since the user may not care about the type and/or source values, we populate them with default values. It's hard to get into a lot of detail on all of this because we are simply referencing external CNCF libraries and specifications.

> If not specified, the CloudEvent `event_source` attribute is set to `"/parliament/function"`, and the `event_type` attribute is set to `"parliament.response"`.

Author: Above comment removes mention of them anyway, probably no need to mention if it's just a default setting and we're linking to the CloudEvent spec
